<div style="background-color: #3FFFF; padding: 30px;"> <div style="text-align: center;">
    <h1 style="font-family: Arial, sans-serif; font-size: 30px; font-weight: bold; text-align: center; margin: 20px 0;">Customer Segmentation using RFM Analysis</h1>
</div> 

<img src="https://rfmcube.com/wp-content/uploads/2021/07/1_HiwX6vul8c4PBEueq3yBMw.png" alt="Hero image" style="width: 100%;">

## What is RFM Analysis?
RFM (Recency, Frequency, Monetary) analysis is a marketing technique used to quantitatively rank and group customers based on their transactional behavior. The goal is to identify the most valuable customers, understand their behaviors, and tailor marketing strategies accordingly.

### Components of RFM:
- Recency (R): How recently a customer made a purchase.
- Frequency (F): How often a customer makes a purchase.
- Monetary (M): How much money a customer spends on purchases.

By scoring customers on these three dimensions, businesses can create customer segments for targeted marketing.

### Importance of RFM in Machine Learning:

- Customer Segmentation: Helps in identifying customer segments for personalized marketing.
- Customer Retention: Identifies at-risk customers and potential loyal customers.
- Profitability Analysis: Helps in recognizing high-value customers.
- Targeted Campaigns: Optimizes marketing efforts and budgets.

### Problem Statement

Utilize RFM analysis to segment customers and drive data-informed marketing strategies for enhanced business outcomes.

#### Step 1: Importing Libraries

In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

from warnings import filterwarnings
filterwarnings('ignore')

#### Step 2: Data Collection

In [3]:
df = pd.read_csv('data.csv', encoding="ISO-8859-1")
df.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

      InvoiceDate  UnitPrice  CustomerID         Country  
0  12/1/2010 8:26       2.55     17850.0  United Kingdom  
1  12/1/2010 8:26       3.39     17850.0  United Kingdom  
2  12/1/2010 8:26       2.75     17850.0  United Kingdom  
3  12/1/2010 8:26       3.39     17850.0  United Kingdom  
4  12/1/2010 8:26       3.39     17850.0  United Kingdom

In [4]:
df1 = df.copy(deep = True)

#### Understanding about the dataset

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    541909 non-null  object 
 1   StockCode    541909 non-null  object 
 2   Description  540455 non-null  object 
 3   Quantity     541909 non-null  int64  
 4   InvoiceDate  541909 non-null  object 
 5   UnitPrice    541909 non-null  float64
 6   CustomerID   406829 non-null  float64
 7   Country      541909 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 33.1+ MB


In [6]:
df.shape

(541909, 8)

In [7]:
df.columns

Index(['InvoiceNo', 'StockCode', 'Description', 'Quantity', 'InvoiceDate',
       'UnitPrice', 'CustomerID', 'Country'],
      dtype='object')

In [8]:
df.describe().T

count          mean          std       min       25%       50%  \
Quantity    541909.0      9.552250   218.081158 -80995.00      1.00      3.00   
UnitPrice   541909.0      4.611114    96.759853 -11062.06      1.25      2.08   
CustomerID  406829.0  15287.690570  1713.600303  12346.00  13953.00  15152.00   

                 75%      max  
Quantity       10.00  80995.0  
UnitPrice       4.13  38970.0  
CustomerID  16791.00  18287.0

In [9]:
df.isnull().sum()

InvoiceNo           0
StockCode           0
Description      1454
Quantity            0
InvoiceDate         0
UnitPrice           0
CustomerID     135080
Country             0
dtype: int64

#### We can see that the Quantity is negative and Unit Price is negative, So we are droping those records which are less than equal too 0

In [10]:
new_df = df[(df.Quantity >0) & (df.UnitPrice >0)]

In [11]:
f'There are totaly {541909 - 530104} transactions which were having negative quantity or negative unit price'

'There are totaly 11805 transactions which were having negative quantity or negative unit price'

In [12]:
new_df.isnull().sum()

InvoiceNo           0
StockCode           0
Description         0
Quantity            0
InvoiceDate         0
UnitPrice           0
CustomerID     132220
Country             0
dtype: int64

#### We can see 132220 customer information is not present so we are going to drop those details.

In [13]:
new_df1 = new_df.dropna()
new_df1

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

            InvoiceDate  UnitPrice  CustomerID         Country  
0        12/1/2010 8:26       2.55     17850.0  United Kingdom  
1        12/1/2010 8:26       3.39     17850.0  United Kingdom  
2        12/1/2010 8:26       2.75     17850.0  United Kingdom  
3        12/1/2010 8:26       3.39     17850.0  United Kingdom  
4        12/1/2010 8:26       3.39     17850.0  United Kingdom  
...                 ...        ...         ...             ...  
541904  12/9/2011 12:50       0.85     12680.0          France  
541905  12/9/2011 12:50       2.10     12680.0          France  
541906  12/9/2011 12:50       4.15     12680.0          France  
541907  12/9/2011 12:50       4.15     12680.0          France  
541908  12/9/2011 12:50       4.95     12680.0          France  

[397884 rows x 8 columns]

#### Now will investigate where are our customers from which Country

In [14]:
new_df1.Country.value_counts()

Country
United Kingdom          354321
Germany                   9040
France                    8341
EIRE                      7236
Spain                     2484
Netherlands               2359
Belgium                   2031
Switzerland               1841
Portugal                  1462
Australia                 1182
Norway                    1071
Italy                      758
Channel Islands            748
Finland                    685
Cyprus                     614
Sweden                     451
Austria                    398
Denmark                    380
Poland                     330
Japan                      321
Israel                     248
Unspecified                244
Singapore                  222
Iceland                    182
USA                        179
Canada                     151
Greece                     145
Malta                      112
United Arab Emirates        68
European Community          60
RSA                         57
Lebanon                     45


In [15]:
new_df1.Country.value_counts(normalize = True)

Country
United Kingdom          0.890513
Germany                 0.022720
France                  0.020963
EIRE                    0.018186
Spain                   0.006243
Netherlands             0.005929
Belgium                 0.005105
Switzerland             0.004627
Portugal                0.003674
Australia               0.002971
Norway                  0.002692
Italy                   0.001905
Channel Islands         0.001880
Finland                 0.001722
Cyprus                  0.001543
Sweden                  0.001133
Austria                 0.001000
Denmark                 0.000955
Poland                  0.000829
Japan                   0.000807
Israel                  0.000623
Unspecified             0.000613
Singapore               0.000558
Iceland                 0.000457
USA                     0.000450
Canada                  0.000380
Greece                  0.000364
Malta                   0.000281
United Arab Emirates    0.000171
European Community      0.000151
RS

#### We can see that alomost 90% of our customers are from United Kingdom (UK) so will do our analysis only for UK

In [16]:
df_uk = new_df1[new_df1.Country == 'United Kingdom']
df_uk

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541889    581585     22466       FAIRY TALE COTTAGE NIGHT LIGHT        12   
541890    581586     22061  LARGE CAKE STAND  HANGING STRAWBERY         8   
541891    581586     23275     SET OF 3 HANGING OWLS OLLIE BEAK        24   
541892    581586     21217        RED RETROSPOT ROUND CAKE TINS        24   
541893    581586     20685                DOORMAT RED RETROSPOT        10   

            InvoiceDate  UnitPrice  CustomerID         Country  
0        12/1/2010 8:26       2.55     17850.0  United Kingdom  
1        12/1/2010 8:26       3.39     17850.0  United Kingdom  
2        12/1/2010 8:26       2.75     17850.0  United Kingdom  
3        12/1/2010 8:26       3.39     17850.0  United Kingdom  
4        12/1/2010 8:26       3.39     17850.0  United Kingdom  
...                 ...        ...         ...             ...  
541889  12/9/2011 12:31       1.95     15804.0  United Kingdom  
541890  12/9/2011 12:49       2.95     13113.0  United Kingdom  
541891  12/9/2011 12:49       1.25     13113.0  United Kingdom  
541892  12/9/2011 12:49       8.95     13113.0  United Kingdom  
541893  12/9/2011 12:49       7.08     13113.0  United Kingdom  

[354321 rows x 8 columns]

In [17]:
df_uk.reset_index(drop=True,inplace = True)

In [18]:
df_uk

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
354316    581585     22466       FAIRY TALE COTTAGE NIGHT LIGHT        12   
354317    581586     22061  LARGE CAKE STAND  HANGING STRAWBERY         8   
354318    581586     23275     SET OF 3 HANGING OWLS OLLIE BEAK        24   
354319    581586     21217        RED RETROSPOT ROUND CAKE TINS        24   
354320    581586     20685                DOORMAT RED RETROSPOT        10   

            InvoiceDate  UnitPrice  CustomerID         Country  
0        12/1/2010 8:26       2.55     17850.0  United Kingdom  
1        12/1/2010 8:26       3.39     17850.0  United Kingdom  
2        12/1/2010 8:26       2.75     17850.0  United Kingdom  
3        12/1/2010 8:26       3.39     17850.0  United Kingdom  
4        12/1/2010 8:26       3.39     17850.0  United Kingdom  
...                 ...        ...         ...             ...  
354316  12/9/2011 12:31       1.95     15804.0  United Kingdom  
354317  12/9/2011 12:49       2.95     13113.0  United Kingdom  
354318  12/9/2011 12:49       1.25     13113.0  United Kingdom  
354319  12/9/2011 12:49       8.95     13113.0  United Kingdom  
354320  12/9/2011 12:49       7.08     13113.0  United Kingdom  

[354321 rows x 8 columns]

#### Now will Start Making our customer segmentation analysis based on 354321 customers transaction

#### Droping unwanted features.

In [19]:
final_df = df_uk.drop(columns = ['InvoiceNo','StockCode','Description','Country'])

In [20]:
final_df.head()

Quantity     InvoiceDate  UnitPrice  CustomerID
0         6  12/1/2010 8:26       2.55     17850.0
1         6  12/1/2010 8:26       3.39     17850.0
2         8  12/1/2010 8:26       2.75     17850.0
3         6  12/1/2010 8:26       3.39     17850.0
4         6  12/1/2010 8:26       3.39     17850.0

In [21]:
final_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 354321 entries, 0 to 354320
Data columns (total 4 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   Quantity     354321 non-null  int64  
 1   InvoiceDate  354321 non-null  object 
 2   UnitPrice    354321 non-null  float64
 3   CustomerID   354321 non-null  float64
dtypes: float64(2), int64(1), object(1)
memory usage: 10.8+ MB


#### Step 2: Data Preparation
Calculate Recency, Frequency, and Monetary values.

In [22]:
from datetime import date

In [23]:
date.today()

datetime.date(2025, 2, 19)

In [24]:
import datetime

In [25]:
# Reference date (today's date for the analysis)
reference_date = datetime.datetime.now()
reference_date

datetime.datetime(2025, 2, 19, 18, 23, 22, 536761)

In [26]:
final_df['Transaction_Amount'] = final_df.Quantity * final_df.UnitPrice
final_df.head(2)

Quantity     InvoiceDate  UnitPrice  CustomerID  Transaction_Amount
0         6  12/1/2010 8:26       2.55     17850.0               15.30
1         6  12/1/2010 8:26       3.39     17850.0               20.34

In [27]:
final_df['InvoiceDate'] = pd.to_datetime(final_df.InvoiceDate)

In [28]:
# Aggregate data by CustomerID
rfm_data = final_df.groupby('CustomerID').agg({
    'InvoiceDate': lambda x: (reference_date - x.max()).days,
    'CustomerID': 'count',
    'Transaction_Amount': 'sum'
}).rename(columns={
    'TransactionDate': 'Recency',
    'CustomerID': 'Frequency',
    'Transaction_Amount': 'Monetary'
})

rfm_data.head() 

InvoiceDate  Frequency  Monetary
CustomerID                                  
12346.0            5146          1  77183.60
12747.0            4823        103   4196.01
12748.0            4821       4595  33719.73
12749.0            4824        199   4090.88
12820.0            4824         59    942.34